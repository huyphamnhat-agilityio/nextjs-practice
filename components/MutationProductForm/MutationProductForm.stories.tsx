import type { Meta, StoryObj } from "@storybook/react";
import { useArgs } from "@storybook/preview-api";

// Components
import MutationProductForm from ".";

const meta: Meta<typeof MutationProductForm> = {
  title: "Components/MutationProductForm",
  component: MutationProductForm,
  argTypes: {
    data: {
      description: "The product data of the form",
    },
    isOpen: {
      description: "Whether the form is open or not",
    },
    onClose: {
      description: "Callback function to close the form",
    },
    onOpen: {
      description: "Callback function to open the form",
    },
    onOpenChange: {
      description: "Callback function to handle form opening/closing changes",
    },
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof MutationProductForm>;

export const Primary: Story = {
  args: {
    isOpen: true,
    onOpen: () => {},
    onOpenChange: () => {},
  },
  render: function Render(args) {
    const [, updateArgs] = useArgs<typeof args>();

    const onClose = () => {
      updateArgs({ isOpen: false });
    };

    return <MutationProductForm {...args} onClose={onClose} />;
  },
};
