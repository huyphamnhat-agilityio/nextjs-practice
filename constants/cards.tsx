// Components
import { HeroCardProps } from "@/components/common/HeroCard";
import {
  BlackBoardIcon,
  MortarboardIcon,
  TelescopeIcon,
} from "@/components/common/Icons";

export const HERO_CARD_LIST: Array<HeroCardProps> = [
  {
    icon: <BlackBoardIcon />,
    iconBackgroundColor: "primary",
    title: "Expert instruction",
    description:
      "The gradual accumulation of information about atomic and small-scale behaviour...",
  },
  {
    icon: <TelescopeIcon />,
    iconBackgroundColor: "secondary",
    title: "Training Courses",
    description:
      "The gradual accumulation of information about atomic and small-scale behaviour...",
  },
  {
    icon: <MortarboardIcon />,
    iconBackgroundColor: "default",
    title: "Lifetime access",
    description:
      "The gradual accumulation of information about atomic and small-scale behaviour...",
  },
];
