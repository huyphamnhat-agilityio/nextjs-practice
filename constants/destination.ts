export const DESTINATION = {
  HOME: "/",
  PRODUCT: "/products",
} as const;
