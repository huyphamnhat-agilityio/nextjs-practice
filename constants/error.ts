export const ERROR_MESSAGES = {
  400: "The request could not be understood by the server due to malformed syntax. Please check your input and try again.",
  401: "You are not authorized to access this resource. Please log in with valid credentials and try again.",
  403: "You do not have permission to access this resource. Please contact the administrator if you believe this is an error.",
  404: "The requested resource could not be found on this server. Please check the URL and try again.",
  405: "The HTTP method used is not allowed for this resource. Please refer to the API documentation for the correct method.",
  406: "The requested resource cannot generate content acceptable according to the Accept headers sent in the request. Please modify your request and try again.",
} as const;
