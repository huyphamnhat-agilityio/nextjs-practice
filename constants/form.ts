export const FORM_STATUS = {
  SUCCESS: "success",
  ERROR: "error",
} as const;
