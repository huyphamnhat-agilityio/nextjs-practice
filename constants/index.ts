export * from "./nav-list";
export * from "./cards";
export * from "./resources";
export * from "./pagination";
export * from "./images";
export * from "./message";
export * from "./destination";
export * from "./tags";
export * from "./toast";
