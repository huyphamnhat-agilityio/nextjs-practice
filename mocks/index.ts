export * from "./category";
export * from "./product";
export * from "./fetch";
export * from "./avatar";
export * from "./image";
