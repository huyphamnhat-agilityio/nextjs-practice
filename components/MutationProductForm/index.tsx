"use client";
import React, { useCallback, useState } from "react";
import toast from "react-hot-toast";
import { Control, Controller, FieldErrors, useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import {
  Image as NextUIImage,
  Modal,
  ModalBody,
  ModalContent,
  ModalHeader,
  SelectItem,
  ModalFooter,
  Spinner,
} from "@nextui-org/react";

// Types
import { FormState, Product, ProductForm } from "@/types";

// Schemas
import { ProductFormSchema } from "@/schemas";

// Mocks
import { MOCK_CATEGORIES, PLACEHOLDER_COURSE_IMAGE } from "@/mocks";

// Components
import { Button, ClearIcon, Input, Select, Textarea } from "../common";

// Services
import { mutateProduct, uploadAndGetImageUrl } from "@/lib";

// Constants
import { PRODUCT_MESSAGES } from "@/constants";

// Utils
import { convertImageToBase64 } from "@/utils";

export type ProductFormBodyProps = {
  selectedImage: File;
  control: Control<ProductForm, any>;
  handleSelectImage: (e: React.ChangeEvent<HTMLInputElement>) => void;
  handleClearImage: () => void;
  errors: FieldErrors<ProductForm>;
  isValid: boolean;
  isPending: boolean;
  isDirty: boolean;
  onClose: () => void;
  coverImageUrl: string;
};
const ProductFormBody = ({
  selectedImage,
  control,
  handleSelectImage,
  handleClearImage,
  errors,
  isValid,
  isPending,
  isDirty,
  onClose,
  coverImageUrl,
}: ProductFormBodyProps) => {
  return (
    <>
      <ModalBody className="flex flex-col gap-6 relative">
        {(!!selectedImage || !!coverImageUrl) && (
          <Button
            aria-label="clear-image"
            className="absolute z-50 top-2 right-6  p-2 text-foreground-500 bg-transparent rounded-full hover:bg-default-100 active:bg-default-200 tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
            isIconOnly
            onClick={handleClearImage}
          >
            <ClearIcon />
          </Button>
        )}
        <label
          aria-label="cover course image url"
          className={`self-center ${isPending && "opacity-50"}`}
          aria-disabled={isPending}
        >
          <NextUIImage
            width={448}
            height={256}
            src={
              selectedImage
                ? URL.createObjectURL(selectedImage)
                : coverImageUrl
                  ? coverImageUrl
                  : PLACEHOLDER_COURSE_IMAGE
            }
            className="object-cover"
            alt=""
          />

          <Controller
            control={control}
            name="coverImage"
            render={({ field: { onChange, onBlur } }) => (
              <Input
                data-testid="cover-image"
                name="coverImage"
                type="file"
                className="hidden"
                onBlur={onBlur}
                onChange={(e) => {
                  if (e.target.files.length > 0) {
                    handleSelectImage(e);
                    onChange(e.target.files[0]);
                  }
                }}
              />
            )}
          />
        </label>
        {errors?.coverImage && (
          <p className="text-base text-danger">{errors.coverImage.message}</p>
        )}

        <Controller
          control={control}
          name="category"
          render={({ field: { onChange, onBlur, value } }) => (
            <Select
              label="Category"
              name="category"
              labelPlacement="outside"
              placeholder="Choose category..."
              value={value}
              selectedKeys={[`${value}`]}
              isDisabled={isPending}
              onChange={(e) => onChange(e.target.value)}
              onBlur={onBlur}
            >
              {MOCK_CATEGORIES.map((category) => (
                <SelectItem key={category.name}>{category.name}</SelectItem>
              ))}
            </Select>
          )}
        />

        <Controller
          control={control}
          name="title"
          render={({ field: { onChange, onBlur, value } }) => (
            <>
              <Input
                label="Title"
                name="title"
                labelPlacement="outside"
                placeholder="Enter course title..."
                isDisabled={isPending}
                isInvalid={!!errors?.title}
                errorMessage={errors?.title?.message}
                value={value}
                onChange={onChange}
                onBlur={onBlur}
              />
            </>
          )}
        />

        <Controller
          control={control}
          name="description"
          render={({ field }) => (
            <Textarea
              {...field}
              label="Description"
              name="description"
              labelPlacement="outside"
              placeholder="Enter course description..."
              isInvalid={!!errors?.description}
              errorMessage={errors?.description?.message}
              isDisabled={isPending}
            />
          )}
        />

        <Controller
          control={control}
          name="sales"
          render={({ field: { onChange, onBlur, value } }) => (
            <Input
              label="Sales"
              name="sales"
              labelPlacement="outside"
              placeholder="Enter sale amount..."
              type="number"
              step="any"
              value={value?.toString()}
              isDisabled={isPending}
              isInvalid={!!errors?.sales}
              errorMessage={errors?.sales?.message}
              onBlur={onBlur}
              onChange={(e) => {
                onChange(
                  !isNaN(e.target.valueAsNumber) ? e.target.valueAsNumber : 0,
                );
              }}
              onKeyDown={(e: React.KeyboardEvent) =>
                ["e", "E", "+", "-", "."].includes(e.key) && e.preventDefault()
              }
            />
          )}
        />
        <Controller
          control={control}
          name="originalPrice"
          render={({ field: { onChange, onBlur, value } }) => (
            <Input
              label="Original Price"
              name="originalPrice"
              labelPlacement="outside"
              placeholder="Enter original price..."
              type="number"
              step="any"
              value={value?.toString()}
              isDisabled={isPending}
              isInvalid={!!errors?.originalPrice}
              errorMessage={errors?.originalPrice?.message}
              onBlur={onBlur}
              onChange={(e) => {
                onChange(
                  !isNaN(e.target.valueAsNumber) ? e.target.valueAsNumber : 0,
                );
              }}
            />
          )}
        />

        <Controller
          control={control}
          name="salePrice"
          render={({ field: { onChange, onBlur, value } }) => (
            <Input
              label="Sale Price"
              name="salePrice"
              labelPlacement="outside"
              placeholder="Enter sale price..."
              type="number"
              step="any"
              value={value?.toString()}
              isDisabled={isPending}
              isInvalid={!!errors?.salePrice}
              errorMessage={errors?.salePrice?.message}
              onBlur={onBlur}
              onChange={(e) => {
                onChange(
                  !isNaN(e.target.valueAsNumber) ? e.target.valueAsNumber : 0,
                );
              }}
            />
          )}
        />

        <Controller
          control={control}
          name="rate"
          render={({ field: { onChange, onBlur, value } }) => (
            <Input
              label="Rate"
              name="rate"
              labelPlacement="outside"
              placeholder="Enter rate..."
              type="number"
              inputMode="decimal"
              step="any"
              value={value?.toString()}
              isDisabled={isPending}
              isInvalid={!!errors?.rate}
              errorMessage={errors?.rate?.message}
              onBlur={onBlur}
              onChange={(e) => {
                onChange(
                  !isNaN(e.target.valueAsNumber) ? e.target.valueAsNumber : 0,
                );
              }}
            />
          )}
        />

        <Controller
          control={control}
          name="isFavorited"
          render={({ field: { value } }) => (
            <Input
              name="isFavorited"
              className="hidden"
              value={value.toString()}
            />
          )}
        />

        <Controller
          control={control}
          name="id"
          render={({ field }) => (
            <Input {...field} name="id" className="hidden" />
          )}
        />

        <Controller
          control={control}
          name="createdAt"
          render={({ field: { value } }) => (
            <Input
              name="createdAt"
              className="hidden"
              value={value.toString()}
            />
          )}
        />

        <Controller
          control={control}
          name="coverImageUrl"
          render={({ field }) => (
            <Input {...field} name="coverImageUrl" className="hidden" />
          )}
        />
      </ModalBody>
      <ModalFooter>
        <Button
          color="default"
          variant="flat"
          className="bg-foreground-100 text-white"
          isDisabled={isPending}
          onPress={onClose}
        >
          Cancel
        </Button>
        <Button
          color="primary"
          className="text-white"
          type="submit"
          spinner={<Spinner size="sm" color="white" />}
          isLoading={isPending}
          isDisabled={!isValid || isPending || !isDirty}
        >
          Submit
        </Button>
      </ModalFooter>
    </>
  );
};

export type MutationProductFormProps = {
  isOpen: boolean;
  onOpen: () => void;
  onOpenChange: () => void;
  onClose: () => void;
  data?: Product;
};

const MutationProductForm = ({
  isOpen,
  onClose,
  data,
}: MutationProductFormProps) => {
  const initialState: FormState<ProductForm> = {};

  const {
    control,
    formState: { errors, isValid, isDirty },
    reset,
    handleSubmit,
    setError,
    getValues,
    setValue,
  } = useForm<ProductForm>({
    mode: "all",
    defaultValues: {
      id: data?.id ?? "",
      category: data?.category ?? "",
      title: data?.title ?? "",
      description: data?.description ?? "",
      sales: data?.sales ?? 0,
      originalPrice: data?.originalPrice ?? 0,
      salePrice: data?.salePrice ?? 0,
      rate: data?.rate ?? 0,
      coverImageUrl: data?.coverImageUrl ?? "",
      coverImage: undefined,
      isFavorited: data?.isFavorited ?? 0,
      createdAt: data?.createdAt ?? 0,
    },
    resolver: zodResolver(ProductFormSchema),
  });

  const [selectedImage, setSelectedImage] = useState<File>();

  const [isPending, setIsPending] = useState(false);

  const handleSelectImage = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setSelectedImage(e.target.files[0]);
    },
    [],
  );

  const handleCloseModal = useCallback(() => {
    reset();
    setSelectedImage(undefined);
    onClose();
  }, [onClose, reset]);

  const handleOnOpenChange = useCallback(
    (isOpen: boolean) => {
      if (!isOpen) handleCloseModal();
    },
    [handleCloseModal],
  );

  const handleClearImage = useCallback(() => {
    selectedImage?.size > 0 && setSelectedImage(undefined);

    getValues("coverImageUrl") &&
      setValue("coverImageUrl", "", { shouldDirty: true });

    getValues("coverImage") &&
      setValue("coverImage", undefined, { shouldDirty: true });
  }, [getValues, selectedImage, setValue]);

  const onSubmit = useCallback(
    async (data: ProductForm) => {
      try {
        setIsPending(true);

        let imageUrl = "";

        try {
          if (data.coverImage) {
            const imageBase64 = await convertImageToBase64(data.coverImage);

            imageUrl = await uploadAndGetImageUrl(imageBase64);
          }
        } catch (error) {
          setError("coverImage", {
            message: PRODUCT_MESSAGES.ERROR.UPLOAD_IMAGE,
          });
          return;
        }

        const { coverImage, coverImageUrl, createdAt, ...rest } = data;

        const product: Product = {
          ...rest,
          coverImageUrl: imageUrl || coverImageUrl,
          createdAt: data.id ? data.createdAt : Date.now(),
        };

        const response = await mutateProduct(product);

        if (response) {
          toast.success(
            data.id
              ? PRODUCT_MESSAGES.SUCCESS.UPDATE
              : PRODUCT_MESSAGES.SUCCESS.CREATE,
          );
          handleCloseModal();
        }
      } catch (error) {
        toast.error(
          data.id
            ? PRODUCT_MESSAGES.ERROR.UPDATE
            : PRODUCT_MESSAGES.ERROR.CREATE,
        );
      } finally {
        setIsPending(false);
      }
    },
    [handleCloseModal, setError],
  );

  return (
    <Modal
      isOpen={isOpen}
      onOpenChange={handleOnOpenChange}
      scrollBehavior="outside"
    >
      <ModalContent>
        <ModalHeader className="text-2xl">
          {data?.id ? "Edit course" : "Add course"}
        </ModalHeader>
        <form onSubmit={handleSubmit(onSubmit)}>
          <ProductFormBody
            control={control}
            errors={errors}
            handleSelectImage={handleSelectImage}
            handleClearImage={handleClearImage}
            isValid={isValid}
            isPending={isPending}
            isDirty={isDirty}
            selectedImage={selectedImage}
            coverImageUrl={getValues("coverImageUrl")}
            onClose={handleCloseModal}
          />
        </form>
      </ModalContent>
    </Modal>
  );
};

export default MutationProductForm;
