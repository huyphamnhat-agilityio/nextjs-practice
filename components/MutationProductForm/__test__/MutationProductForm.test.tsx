import { render, screen } from "@testing-library/react";
import { useFormStatus } from "react-dom";
import { usePathname, useSearchParams } from "next/navigation";
import userEvent from "@testing-library/user-event";
import MutationProductForm, { MutationProductFormProps } from "..";

// Mocks
import { MOCK_PRODUCTS, PLACEHOLDER_PRODUCT_FORM_DATA } from "@/mocks";
import { FORM_MESSAGES } from "@/constants";

jest.mock("react-dom", () => ({
  ...jest.requireActual("react-dom"),
  useFormStatus: jest.fn(),
  useFormState: () => [{}, () => {}],
}));

jest.mock("next/navigation", () => ({
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
}));

describe("MutationProductForm test cases", () => {
  beforeAll(() => {
    window.addEventListener("submit", (e) => {
      e.preventDefault();
    });
  });

  beforeEach(() => {
    (useFormStatus as jest.Mock).mockReturnValue({ pending: false });
    (usePathname as jest.Mock).mockReturnValue("/");
    (useSearchParams as jest.Mock).mockReturnValue(new URLSearchParams());
  });

  const setup = (props: MutationProductFormProps) =>
    render(<MutationProductForm {...props} />);

  const mockProps: MutationProductFormProps = {
    isOpen: true,
    onClose: jest.fn(),
    onOpen: jest.fn(),
    onOpenChange: jest.fn(),
    data: MOCK_PRODUCTS[0],
  };

  it("should renders correctly with no data", () => {
    const { asFragment } = setup({
      ...mockProps,
      data: undefined,
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("should renders correctly with existing data", () => {
    const { asFragment } = setup(mockProps);

    expect(asFragment()).toMatchSnapshot();
  });

  it("disables the submit button when pending", () => {
    (useFormStatus as jest.Mock).mockReturnValue({ pending: true });

    setup(mockProps);

    const submitBtn = screen.getByRole("button", {
      name: /submit/i,
    });

    expect(submitBtn).toBeDisabled();
  });

  it("should be able to submit when entering valid data and clear data when close form", async () => {
    setup(mockProps);

    const titleInput = screen.getByRole("textbox", {
      name: /title title/i,
    });

    await userEvent.type(titleInput, "Updated Title");

    const submitBtn = screen.getByRole("button", {
      name: /submit/i,
    });

    expect(submitBtn).not.toBeDisabled();

    const closeBtn = screen.getByRole("button", {
      name: /cancel/i,
    });

    await userEvent.click(closeBtn);
  });

  it("should be able to prevent entering invalid data", async () => {
    setup(mockProps);

    const titleInput = screen.getByRole("textbox", {
      name: /title title/i,
    });

    await userEvent.type(titleInput, "Updated Title");

    const salesInput = screen.getByRole("spinbutton", {
      name: /sales sales/i,
    });

    await userEvent.type(salesInput, "-1");

    const originalPriceInput = screen.getByRole("spinbutton", {
      name: /original price original price/i,
    });

    await userEvent.type(originalPriceInput, "-1");

    const salePriceInput = screen.getByRole("spinbutton", {
      name: /sale price sale price/i,
    });

    await userEvent.type(salePriceInput, "-1");

    const rateInput = screen.getByRole("spinbutton", {
      name: /rate rate/i,
    });

    await userEvent.type(rateInput, "-1");

    const categorySelectList = screen.getByRole("button", {
      name: /training courses category/i,
      hidden: true,
    });

    await userEvent.click(categorySelectList);

    const categoryOption = screen.getByRole("option", {
      name: /books library/i,
    });

    await userEvent.click(categoryOption);

    expect(titleInput).toHaveValue(`${mockProps.data.title}Updated Title`);
    expect(categorySelectList).toHaveValue("Books Library");
    expect(salesInput).toHaveValue(1);
    expect(salePriceInput).toHaveValue(1);
    expect(originalPriceInput).toHaveValue(1);
    expect(rateInput).toHaveValue(1);
  });

  it("should display error message when typing invalid data", async () => {
    window.URL.createObjectURL = jest.fn();

    const { asFragment } = setup({
      ...mockProps,
      data: {
        ...PLACEHOLDER_PRODUCT_FORM_DATA,
        title: "a",
        sales: -1,
        originalPrice: -1,
        salePrice: -1,
        rate: -1,
      },
    });

    const file = new File([""], "mock.svg", { type: "image/svg+xml" });

    const imgInput = screen.getByTestId("cover-image");

    await userEvent.upload(imgInput, file);

    await userEvent.tab();

    await userEvent.tab();

    await userEvent.tab();

    await userEvent.tab();

    await userEvent.tab();

    await userEvent.tab();

    await userEvent.tab();

    await userEvent.tab();

    const imgError = screen.getByText(
      new RegExp(FORM_MESSAGES.PRODUCT.COVER_IMAGE.ACCEPTED_FORMATS, "i"),
    );

    const titleError = screen.getByText(
      new RegExp(FORM_MESSAGES.PRODUCT.TITLE.MIN, "i"),
    );

    const salesError = screen.getByText(
      new RegExp(FORM_MESSAGES.PRODUCT.SALES.MIN, "i"),
    );

    const originalPriceError = screen.getByText(
      new RegExp(FORM_MESSAGES.PRODUCT.ORIGINAL_PRICE.MIN, "i"),
    );

    const salePriceError = screen.getByText(
      new RegExp(FORM_MESSAGES.PRODUCT.SALE_PRICE.MIN, "i"),
    );

    const rateError = screen.getByText(
      new RegExp(FORM_MESSAGES.PRODUCT.RATE.MIN, "i"),
    );

    expect(imgError.textContent).toEqual(
      FORM_MESSAGES.PRODUCT.COVER_IMAGE.ACCEPTED_FORMATS,
    );

    expect(titleError.textContent).toEqual(FORM_MESSAGES.PRODUCT.TITLE.MIN);

    expect(salesError.textContent).toEqual(FORM_MESSAGES.PRODUCT.SALES.MIN);

    expect(originalPriceError.textContent).toEqual(
      FORM_MESSAGES.PRODUCT.ORIGINAL_PRICE.MIN,
    );

    expect(salePriceError.textContent).toEqual(
      FORM_MESSAGES.PRODUCT.SALE_PRICE.MIN,
    );

    expect(rateError.textContent).toEqual(FORM_MESSAGES.PRODUCT.RATE.MIN);
  });
});
