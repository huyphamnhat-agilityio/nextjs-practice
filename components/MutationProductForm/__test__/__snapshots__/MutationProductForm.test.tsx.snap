// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MutationProductForm test cases should renders correctly with existing data 1`] = `<DocumentFragment />`;

exports[`MutationProductForm test cases should renders correctly with no data 1`] = `<DocumentFragment />`;
